Add json,yaml output format support to kubectl create, kubectl apply

juanvallejo opened the pull request:
Fixes: https://github.com/kubernetes/kubernetes/issues/37390

**Release note**:
```release-note
Added support for printing in all supported `--output` formats to `kubectl create ...` and `kubectl apply ...`
```

This patch adds the ability to specify an output format other than
"name" to `kubectl create ...`. It can be used in conjunction with the
`--dry-run` option. Converts unstructured objects into known types in
order to support all `--output` values.

The patch prints `*resource.Info`s returned by the server. If a resource does not yet exist (and the `--dry-run` option is not set), the resource is created and printed in the specified format.

@kubernetes/cli-review @fabianofranz

k8s-reviewable commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.kubernetes.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/38112)
<!-- Reviewable:end -->

deads2k commented:
Is the output what I sent or what I got back.  I expect it to be what I got back.

juanvallejo commented on pkg/kubectl/cmd/create.go:
```diff
@@ -170,6 +187,20 @@ func RunCreate(f cmdutil.Factory, cmd *cobra.Command, out, errOut io.Writer, opt
+func decodeUnstructuredObject(obj runtime.Object) (runtime.Object, bool) {
+	switch obj.(type) {
+	case *runtime.UnstructuredList, *runtime.Unstructured, *runtime.Unknown:
+		if objBytes, err := runtime.Encode(api.Codecs.LegacyCodec(), obj); err == nil {
```
@deads2k 
> This looks very, very wrong. What are you trying to do?

I'm attempting to convert to a known type, I used [this implementation in filter.go](https://github.com/kubernetes/kubernetes/blob/master/pkg/kubectl/resource_filter.go#L65) as an example for this.

deads2k commented:
> I'm attempting to convert to a known type, I used this implementation in filter.go as an example.

Why would I want to do that?  The whole point of these commands is to be generic if possible.

juanvallejo commented:
@deads2k 
> Why would I want to do that? The whole point of these commands is to be generic if possible.

`--output wide` returned an error such as:
```
error: unknown type &runtime.Unstructured{Object:map[string]interface {}{"status":map[string]interface {}{"lastVersion":1}, "kind":"BuildConfig", "apiVersion":"v1", "metadata":map[string]interface {}{"name":"gitauthtest", "namespace":"default", "selfLink":"/oapi/v1/namespaces/default/buildconfigs/gitauthtest", "uid":"39b2c5a1-b1b3-11e6-9a01-507b9dac96e1", "resourceVersion":"800167", "creationTimestamp":"2016-11-23T19:30:00Z", "labels":map[string]interface {}{"app":"gitauthtest", "template":"gitserver"}, "annotations":map[string]interface {}{"openshift.io/generated-by":"OpenShiftNewApp"}}, "spec":map[string]interface {}{"runPolicy":"Serial", "source":map[string]interface {}{"type":"Git", "git":map[string]interface {}{"uri":"http://gitserver-tokenauth.linux.xip.io/ruby-hello-world"}, "sourceSecret":map[string]interface {}{"name":"builder-token-nbme5"}}, "strategy":map[string]interface {}{"type":"Source", "sourceStrategy":map[string]interface {}{"from":map[string]interface {}{"kind":"ImageStreamTag", "namespace":"openshift", "name":"ruby:latest"}}}, "output":map[string]interface {}{}, "resources":map[string]interface {}{}, "postCommit":map[string]interface {}{}, "nodeSelector":interface {}(nil), "triggers":[]interface {}{}}}}
```

if I did not convert the object here, however another solution in the [downstream PR](https://github.com/openshift/origin/pull/12015) was also to just output something like `Output format not supported` if a printer required object conversion.

deads2k commented:
> --output wide returned an error such as:

I'd rather see the conversion contained in something like the printer that knows what it cares about.  As we move GET to the server, this should fall out of the flow

fabianofranz commented:
@kubernetes/kubectl

juanvallejo commented:
@deads2k 
> I'd rather see the conversion contained in something like the printer that knows what it cares about. As we move GET to the server, this should fall out of the flow

Thanks for the feedback. It looks like `runtime.Unstructurted` objects are already being converted to known types in the `HumanReadablePrinter`'s [printObj](https://github.com/kubernetes/kubernetes/blob/master/pkg/kubectl/resource_printer.go#L2396) method. I went ahead and updated `create.go` so that a printer with an unversioned mapping is used, which allow me to send an unstructured `info.Object` to this method.

ymqytw commented:
@juanvallejo Do you have another PR to fix the similar issue with `kubectl apply`. Ref: https://github.com/kubernetes/kubernetes/pull/34028#issuecomment-262392748

juanvallejo commented:
@ymqytw 
> Do you have another PR to fix the similar issue with kubectl apply

Sorry about that, I had forgotten to address this. Will add as part of this PR. Thanks!

juanvallejo commented:
@ymqytw 
Updated `kubectl apply` to support different output formats as well, PTAL https://github.com/kubernetes/kubernetes/pull/38112/commits/c6658eaee12d99cc07a818460659041846336b5a

juanvallejo commented on pkg/kubectl/cmd/apply.go:
```diff
@@ -251,6 +251,20 @@ func RunApply(f cmdutil.Factory, cmd *cobra.Command, out io.Writer, options *App
 			}
 
 			count++
+			if len(output) > 0 && !shortOutput {
```
@smarterclayton done! extracted this out to `kubectl/cmd/util/printing.go` PTAL

ymqytw commented on pkg/kubectl/cmd/util/printing.go:
```diff
@@ -133,6 +134,25 @@ func PrinterForCommand(cmd *cobra.Command) (kubectl.ResourcePrinter, bool, error
+		if err != nil {
+			return err
+		}
+		return printer.PrintObj(info.Object, out)
```
I think we don't need this line.

juanvallejo replied:
This line is using the printer declared in line [142](https://github.com/kubernetes/kubernetes/pull/38112/files/66f31212ac1d9cd8bc1b6cc5ea3c22b87bcaef0b#diff-6207a4b98b40755e352878f799b7677fR142) although I realize that this entire part could read a bit better. I'll go ahead and rename the `printer` declared inside of the `!generic || printer == nil` block to something a bit more specific, like `nonGenericPrinter`

ymqytw commented:
This PR should have a release note, since it changes the output behavior.

juanvallejo commented:
@ymqytw 
> This PR should have a release note, since it changes the output behavior.

Added a release note

juanvallejo commented:
@fabianofranz or @ymqytw Is this patch okay to test? Thanks!

ymqytw commented:
@k8s-bot ok to test

juanvallejo committed:
Add json,yaml output format support kubectl create

This patch adds the ability to specify an output format other than
"name" to `kubectl create ...`. It can be used in conjunction with the
`--dry-run` option. Converts unstructured objects into known types in
order to support all `--output` values.

juanvallejo commented:
@ymqytw or @deads2k are there any more comments on this patch? Thanks!

deads2k commented:
The description still isn't clear to me. Is this showing me what was submitted or what was actually created (returned from the API).  You definitely need tests.

Also, the printing API needs a significant overhaul.  I'll let @fabianofranz decide what order things need to happen.  I think this reduces overall readability.

juanvallejo commented:
@deads2k 
> The description still isn't clear to me. Is this showing me what was submitted or what was actually created (returned from the API).

Thanks, went ahead and updated the description to clarify this. The printed objects are the ones [returned from the server](https://github.com/juanvallejo/kubernetes/blob/cbe479039b04095f1af606166cbd6d2ad5b0e723/pkg/kubectl/cmd/apply.go#L227-L261). If an info is returned and does not yet exist in the server (and the `--dry-run` option was not also provided with the `--output` flag, then it is [first created](https://github.com/juanvallejo/kubernetes/blob/cbe479039b04095f1af606166cbd6d2ad5b0e723/pkg/kubectl/cmd/apply.go#L243-L253) and then [it is printed](https://github.com/juanvallejo/kubernetes/blob/cbe479039b04095f1af606166cbd6d2ad5b0e723/pkg/kubectl/cmd/apply.go#L256-L258) in the format specified.

juanvallejo commented:
@deads2k 
> Also, the printing API needs a significant overhaul. I'll let @fabianofranz decide what order things need to happen. I think this reduces overall readability.

@fabianofranz I agree that adding another helper to handle printing resources does not feel completely right. The reason why the new helper [PrintResourceInfoForCommand](https://github.com/kubernetes/kubernetes/pull/38112/files#diff-6207a4b98b40755e352878f799b7677fR141) was added in this PR was to handle cases where `--output wide` would cause a panic due to [the way it is currently handled in `kubectl.GetPrinter`](https://github.com/kubernetes/kubernetes/blob/master/pkg/kubectl/resource_printer.go#L144). Perhaps the best solution would be to update that line in `resource_printer.go` to return a `NewHumanReadablePrinter`, or maybe have it actually return an error if it is going to return a `nil` printer and handle returning a `HumanReadablePrinter` with the correct `PrintOptions` in [cmdutil.PrinterForCommand](https://github.com/juanvallejo/kubernetes/blob/cbe479039b04095f1af606166cbd6d2ad5b0e723/pkg/kubectl/cmd/util/printing.go#L109).

Maybe this is something that could be added to the [kubectl umbrella issue for printers](https://github.com/kubernetes/kubernetes/issues/38768)?

fabianofranz commented:
Added https://github.com/kubernetes/kubernetes/issues/38779 to the [printers and describers umbrella](https://github.com/kubernetes/kubernetes/issues/38768). I agree it needs a significant overhaul which we are proposing to do after an analysis of the existing issues, but let's do that in a separate effort.

fabianofranz commented:
Any other comments here?

fabianofranz commented:
@k8s-bot bazel test this
